posts/fantastic-simprocs.md

```diff
@@ -0,0 +1,219 @@
+```
+
+When presented with a left hand side of the form `∃ x, P x`, where `P x` is of the form `_ ∧ ... ∧ _`, `existsAndEq` does the following:
+- Recursively traverse `P x` inside the existential quantifier looking for an equality `x = a` for some `a`.
+- If an equality is found, construct a proof that `∀ x, p x → x = a`.
+- Return the right hand side `P a` together with the proof obtained from the following lemma:
+  ```lean
+  lemma exists_of_imp_eq {α : Sort u} {p : α → Prop} (a : α) (h : ∀ x, p x → x = a) :
+      (∃ x, p x) = p a
+  ```
+
+> Warning: The `existsAndEq` simproc is under active development and the underlying algorithm might change in the future.
+> The above description is only accurate at the time of writing (April 2025).
+
+## Computation
+
+Computations are an integral part of theorem proving, and as such there are many ways to perform them.
+For example, you will find that the `decide` tactic closes all the examples in this subsection.
+There are a few reasons why simprocs are interesting for computation regardless:
+* **`decide` relies on decidability instances**.
+  Not everything one may want to compute is decidable, and not every decidability instance is efficient.
+  In fact, most `Decidable` instances in Lean and Mathlib are very generic, and therefore unspecific and inefficient.
+  Using a simproc gives the opportunity to use a domain-specific algorithm, which is more likely to be efficient and does not rely on the decidability machinery.
+* **`decide` cannot compute the right hand side**, given the left hand side only.
+  `decide` only works on goals that do not contain any metavariable.
+  This rules out using `decide` to discover out what a left hand side is equal to.
+  One would need to write down the right hand side we are looking for in order for `decide` to show that it's equal to the left hand side.
+
+### The `Nat.reduceDvd` simproc
+
+The `Nat.reduceDvd` simproc is designed to take expressions of the form `a | b` where `a, b` are natural number literals, and simplify them to `True` or `False`.
+
+```lean
+example : 3 ∣ 9 := by
+  simp only [Nat.reduceDvd]
+
+example : ¬ 2 ∣ 49 := by
+  simp only [Nat.reduceDvd]
+  -- Remaining goal: `¬ False`
+  simp
+
+example : a ∣ a * b := by
+  simp only [Nat.reduceDvd] --fails
+```
+
+When presented with a left hand side of the form `a ∣ b` where `a` and `b` are natural numbers, `Nat.reduceDvd` does the following:
+- Check that `a` and `b` are numerals.
+- Compute `b % a`.
+- If `b % a` is zero, then return the right hand side `True` together with the proof `Nat.dvd_eq_true_of_mod_eq_zero a b rfl`.
+- If `b % a` isn't zero, then return the right hand side `False` together with the proof `Nat.dvd_eq_false_of_mod_ne_zero a b rfl`.
+
+### The `Finset.Icc_ofNat_ofNat` simproc
+
+The `Finset.Icc_ofNat_ofNat` simproc is designed to take expressions of the form `Finset.Icc a b` where `a` and `b` are numerals, and simplify them to an explicit set.
```

Comment:
It might be worth giving a link to docs for `Finset.Icc`; I'd imagine some readers might not know what it is!